kubespawner/clients.py

```diff
@@ -34,27 +33,36 @@ def shared_client(ClientType, *args, **kwargs):
     """Return a shared kubernetes client instance
     based on the provided arguments.
 
-    A weak reference to the instance is cached,
-    so that concurrent calls to shared_client
-    will all return the same instance until
-    all references to the client are cleared.
+    Cache is one client per running loop per combination of input args.
+
+    Client will be closed when the loop closes.
     """
     kwarg_key = tuple((key, kwargs[key]) for key in sorted(kwargs))
-    cache_key = (ClientType, args, kwarg_key)
-    client = None
-    if cache_key in _client_cache:
-        # resolve cached weakref
-        # client can still be None after this!
-        client = _client_cache[cache_key]()
+    cache_key = (asyncio.get_running_loop(), ClientType, args, kwarg_key)
+    client = _client_cache.get(cache_key, None)
 
     if client is None:
         # Kubernetes client configuration is handled globally and should already
         # be configured from spawner.py or proxy.py via the load_config function
         # prior to a shared_client being instantiated.
         Client = getattr(kubernetes_asyncio.client, ClientType)
         client = Client(*args, **kwargs)
-        # cache weakref so that clients can be garbage collected
-        _client_cache[cache_key] = weakref.ref(client)
+
+        _client_cache[cache_key] = client
+
+        # create a task that will close the client when it is cancelled
+        # relies on JupyterHub's task cleanup at shutdown
+        async def close_client_task():
+            try:
+                async with client.api_client:
+                    while True:
+                        await asyncio.sleep(300)
+            except asyncio.CancelledError:
+                pass
+            finally:
+                _client_cache.pop(cache_key, None)
```

Comment:
Hmm... You are using `with` as a way to ensure that its `__exit__()` is called (which should be `close()`), but you will also end up calling `__enter__`.

I considered something like this, I'm not at all confident about this suggestion though - probably wrong but I don't get the details here yet.

Perhaps is relate to what CancelledError you expect to be thrown and passed etc as well - perhaps you can be explicit about what you mean to catch->pass there?

```suggestion
            try:
                while True:
                    await asyncio.sleep(300)
            except asyncio.CancelledError:
                pass
            finally:
                await client.api_client.close():
                _client_cache.pop(cache_key, None)
```

Reply:
Yes, this is fine, too! `__enter__` only is `return self` so there's no reason to avoid it.

I think the only difference is `with` is the standard Python way to say "clean up after yourself as you see fit", whereas yours explicitly calls `close`, which we know happens to be the cleanup method.

I initially did the `with` because I didn't have the outer try.

Reply:
Ah, the key aspect I appreciate about my method is that we try-catch the CancelledError specifically for asyncio.sleep() now.

Perhaps we can have `async with client.api_client` in the finally statement?

I want to adopt common practices etc, so let's go with what you think is common practice!

Reply:
I guess one could call `__exit__()` as well directly? But that may depend on something being done in `__enter__()` so probably a bad idea also.

Reply:
I think `with` where supported is the most appropriate way to let things clean up after themselves. It should handle errors, etc. appropriately. Then in our own `try/finally` we do our _own_ cleanup (the cache pop).

The reason we catch and suppress the CancelledError is that our task is actually complete at this point, so we don't need to propagate our CancelledError up to anyone checking if we were interrupted.

Reply:
With "our task", do you mean `await asyncio.sleep(300)`? I think I understand things a bit better.

Another idea given that we should stick with `with` is this.

```suggestion
            try:
                while True:
                    await asyncio.sleep(300)
            except asyncio.CancelledError:
                pass
            finally:
                async with client.api_client:
                    pass
                _client_cache.pop(cache_key, None)
```

I'm focused on my own and others comprehension of this code only at this point, not a belief that what we have already is problematic etc.

Reply:
> With "our task", do you mean

I mean `_client_cache.pop`, that's the only resource we have to clean up, so that's the only code _of ours_ in the finally block.

> I'm focused on my own and others comprehension of this code only at this point

absolutely! I think what I have makes the most sense of anything so far, but I certainly want it to be clear.

I read this code:

```python
 async with client.api_client:
    while True:
        await asyncio.sleep(300)
```

as: "hold the api_client resource forever, then let it cleanup when this coroutine is finished, however and whenever that happens".

Putting `with` in the finally block reads like _acquiring_ resources after we're done, which seems odd (though as we  happen to know, acquisition doesn't do anything, so there's no practical difference, other than clarity). If we want to put the client cleanup explicitly in the finally block, I think calling `close` there is better.

Reply:
Thanks for helping me understand this better, it reads more clear to me now after having reasoned about it with you!